[PREGEL] Add Conductor state machine and use it for the Loading state

jvolmer opened the pull request:
Adds a finite state machine to the conductor to improve readability: The behavior of the conductor will now be determined by the conductor state. Before, the conductor only had a enum type `_state`, which had to be checked many times. Now state changes will be defined inside a specific conductor state and can happen when the state runs or gets a message.
This is only the first PR concerning the state machine: It adds only the LoadingState and AnotherState which is just a placeholder for the following state - it will be removed later on. This PR will not break anything, the old enum type `_state` will be deleted once all conductor states are added.

jvolmer commented:
TODO merge parent first

romanatarango commented on arangod/Pregel/Conductor.cpp:
```diff
@@ -1141,3 +1122,15 @@ void Conductor::updateState(ExecutionState state) {
   }
 }
+
+auto Conductor::changeState(conductor::StateType name) -> void {
```
For me, `changeState` sounds as if only the state is changed. However, it is also produced and `run()` is performed. Shouldn't these three things be separated?

jvolmer replied:
The states are designed such that their run() method should be executed as soon as a state-change happened.
Like this the conductor is either running (executes the state->run() method) or waits for a message from a worker which will change the state in the state->receive method and then triggers state->run() on the new state again.
Yes, in this state-machine the states are created just when they are triggered. I chose to do that because the conductor reaches each state only once - except if a recovery happens, but I assume this will not happen too often. That this state is created is a detail which is not relevant for the caller.

markuspf reviewed (changes requested):
Please name header files `.h` not `.hpp` for consistency

jvolmer commented:
Renamed header files and fixed a bug: Inside Loading::receive(Message) we collect messages from all Workers and change the state only when all servers responded. This will be refactored when all state are there, currently it is basically copied from the conductor into the state.

markuspf commented on arangod/Pregel/Conductor/LoadingState.cpp:
```diff
@@ -0,0 +1,58 @@
+
+using namespace arangodb::pregel::conductor;
+
+#define LOG_PREGEL(logId, level)          \
+  LOG_TOPIC(logId, level, Logger::PREGEL) \
+      << "[job " << conductor._executionNumber << "] "
```
Isn't this something that should go into a common header file for Pregel?

jvolmer replied:
ah, good point. It can only be used for the conductor states, so I will put it in the conductor namespace in Conductor/State.h

markuspf replied:
maybe it shouldn't be called `LOG_PREGEL` then, but `LOG_PREGEL_CONDUCTOR`?

markuspf commented on arangod/Pregel/Conductor/LoadingState.cpp:
```diff
@@ -0,0 +1,58 @@
+      << "[job " << conductor._executionNumber << "] "
+
+Loading::Loading(Conductor& conductor) : conductor{conductor} {
+  conductor.updateState(ExecutionState::LOADING);
```
Is this going to be removed again later? This looks like we have 2 different states now.

jvolmer replied:
Yes, definitely, the ExecutionState will be removed when the refactoring is done. Currently we still need it at some places.

markuspf commented on arangod/Pregel/Conductor/LoadingState.cpp:
```diff
@@ -0,0 +1,58 @@
+}
+
+Loading::~Loading() {
+  conductor._timing.loading.finish();
```
How does this destructor interact with conductor's `ExecutionState`?

jvolmer replied:
It does not interact at all. The state is destructed when conductor.changeState is called. In the constructor of the new state, ExecutationState is set to the new state. But we want to get rid of ExecutionState anyway.

markuspf commented on arangod/Pregel/Conductor/LoadingState.cpp:
```diff
@@ -0,0 +1,58 @@
+auto Loading::receive(Message const& message) -> void {
+  if (message.type() != MessageType::GraphLoaded) {
+    LOG_PREGEL("14df4", WARN)
+        << "When loading, we expect a GraphLoaded message, nothing else";
```
Maybe we want to say here which message we received? that might ease debugging quite a bit.

markuspf reviewed (approved):
Some questions
